File: java/src/main/java/it/giannibombelli/wsc2026/booking/application/query/BookingQueryService.java
package it.giannibombelli.wsc2026.booking.application.query;

import it.giannibombelli.wsc2026.booking.domain.booking.Booking;
import it.giannibombelli.wsc2026.booking.domain.booking.BookingId;
import it.giannibombelli.wsc2026.booking.domain.ports.BookingRepository;

import java.util.Optional;

public final class BookingQueryService {
    private final BookingRepository repository;

    public BookingQueryService(BookingRepository repository) {
        this.repository = repository;
    }

    public Optional<BookingDetails> findById(BookingId id) {
        return repository.findById(id)
            .map(this::toDetails);
    }

    private BookingDetails toDetails(Booking booking) {
        return new BookingDetails(
            booking.id().value(),
            booking.description(),
            booking.giftCardId().value()
        );
    }
}

File: typescript/src/booking/application/query/bookingQueryService.ts
// Query service che espone i dati delle prenotazioni all'API.

import { BookingId } from '../../domain/booking/bookingId.js';
import type { BookingRepository } from '../../domain/ports/bookingRepository.js';
import { type BookingDetails } from './bookingDetails.js';

export class BookingQueryService {
  private readonly _repository: BookingRepository;

  constructor(repository: BookingRepository) {
    this._repository = repository;
  }

  findById(id: BookingId): BookingDetails | null {
    const booking = this._repository.findById(id);
    if (booking === null) {
      return null;
    }
    return {
      id: booking.id().value.value,
      description: booking.description(),
      giftCardId: booking.giftCardId().value.value,
    };
  }
}
